cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -621,193 +563,181 @@ public Void apply(AppMetadataStore mds) throws Exception {
-          ApplicationSpecification newAppSpec = new AppSpecificationWithChangedSchedules(appSpec, schedules);
-          mds.updateAppSpec(program.getNamespaceId(), program.getApplicationId(), newAppSpec);
-          return null;
-        }
-      }, apps.get());
+    txExecute(transactional, new TxRunnable() {
+      @Override
+      public void run(DatasetContext context) throws Exception {
+        AppMetadataStore metaStore = getAppMetadataStore(context);
+        ApplicationSpecification appSpec = getAppSpecOrFail(metaStore, program);
+        Map<String, ScheduleSpecification> schedules = Maps.newHashMap(appSpec.getSchedules());
+        String scheduleName = scheduleSpecification.getSchedule().getName();
+        Preconditions.checkArgument(!schedules.containsKey(scheduleName), "Schedule with the name '" +
+          scheduleName + "' already exists.");
+        schedules.put(scheduleSpecification.getSchedule().getName(), scheduleSpecification);
+        ApplicationSpecification newAppSpec = new AppSpecificationWithChangedSchedules(appSpec, schedules);
+        metaStore.updateAppSpec(program.getNamespaceId(), program.getApplicationId(), newAppSpec);
+      }
+    });
   }
 
   @Override
   public void deleteSchedule(final Id.Program program, final String scheduleName) {
-    appsTx.get().executeUnchecked(
-      new TransactionExecutor.Function<AppMetadataStore, Void>() {
-        @Override
-        public Void apply(AppMetadataStore mds) throws Exception {
-          ApplicationSpecification appSpec = getAppSpecOrFail(mds, program);
-          Map<String, ScheduleSpecification> schedules = Maps.newHashMap(appSpec.getSchedules());
-          ScheduleSpecification removed = schedules.remove(scheduleName);
-          if (removed == null) {
-            throw new NoSuchElementException("no such schedule @ account id: " + program.getNamespaceId() +
-                                               ", app id: " + program.getApplication() +
-                                               ", program id: " + program.getId() +
-                                               ", schedule name: " + scheduleName);
-          }
-
-          ApplicationSpecification newAppSpec = new AppSpecificationWithChangedSchedules(appSpec, schedules);
-          mds.updateAppSpec(program.getNamespaceId(), program.getApplicationId(), newAppSpec);
-          return null;
+    txExecute(transactional, new TxRunnable() {
+      @Override
+      public void run(DatasetContext context) throws Exception {
+        AppMetadataStore metaStore = getAppMetadataStore(context);
+        ApplicationSpecification appSpec = getAppSpecOrFail(metaStore, program);
+        Map<String, ScheduleSpecification> schedules = Maps.newHashMap(appSpec.getSchedules());
+        ScheduleSpecification removed = schedules.remove(scheduleName);
+        if (removed == null) {
+          throw new NoSuchElementException("no such schedule @ account id: " + program.getNamespaceId() +
```

Comment:
I know this was wrong before, but can you fix this error message? It talks about account ids when meaning namespace.